OpenRA.Mods.Common/Traits/Armament.cs

```diff
@@ -62,7 +62,28 @@ public class ArmamentInfo : UpgradableTraitInfo, Requires<AttackBaseInfo>
 		[Desc("Use multiple muzzle images if non-zero")]
 		public readonly int MuzzleSplitFacings = 0;
 
+		WeaponInfo weaponInfo = null;
+
+		// Cache initial modified range using sentinal value to mean not cached yet
+		WDist modifiedRange = WDist.MaxValue;
+
 		public override object Create(ActorInitializer init) { return new Armament(init.Self, this); }
+
+		public WeaponInfo GetWeaponInfo(World w)
+		{
+			if (weaponInfo == null)
+				weaponInfo = w.Map.Rules.Weapons[Weapon.ToLowerInvariant()];
+			return weaponInfo;
+		}
+
+		public WDist GetModifiedRange(World w, ActorInfo ai)
```

Comment:
Lazy<> would not work here (or above) since the code needs access to the `Ruleset` and the `ActorInfo`. However, I have a solution similar to `INotifyCreated`.